dali/pipeline/operator/sequence_operator.h

```diff
@@ -0,0 +1,627 @@
+   * Arguments marked as supporting per-frame tensors and with leading `F` in tensor layout are
+   * unfolded (and additionally broadcasted if the channels in the operator's input are unfolded) to
+   * match the operator's input. If the argument is unfolded, each sample of the argument must have
+   * the number of frames equal to either: one (so it can be broadcasted) or to the number of frames
+   * in the corresponding sample of operator's expandable input.
+   *
+   * Tensor inputs not marked per-frame or with no leading frames in the layout are broadcasted to
+   * match the operator's expandable input.
+   *
+   * Assumes ``IsExpandble()`` is true, i.e. there is some expandable input to match the expanded
+   * arguments.
+   */
+  virtual void ExpandArguments(const ArgumentWorkspace &ws) {
+    for (const auto &arg_input : ws) {
+      auto &shared_tvec = arg_input.second.tvec;
+      assert(shared_tvec);
+      ExpandArgument(ws, arg_input.first, *shared_tvec);
+    }
+  }
+
+  /**
+   * @brief Gets the ExpandDesc instance that describes the expandable prefix of the layout and
+   * corresponding prefixes of the shapes in the batch. By default, the sequence shape and
+   * expandable layout are assumed to be the same for the output and the expandable input. If that's
+   * not the case, the method may be overriden to provide different ExpandDesc instance.
+   */
+  virtual const ExpandDesc &GetOutputExpandDesc(const workspace_t<Backend> &ws,
+                                                int output_idx) const {
+    (void)output_idx;
+    return GetInputExpandDesc(ExpandLikeIdx());
+  }
+
+  virtual int GetArgExpandDescIdx(const std::string &arg_name) const {
+    (void)arg_name;
+    return ExpandLikeIdx();
+  }
+
+  const ExpandDesc &GetInputExpandDesc(int input_idx) const {
+    DALI_ENFORCE(0 <= input_idx && static_cast<size_t>(input_idx) < input_expand_desc_.size());
+    return input_expand_desc_[input_idx];
+  }
+
+  int ExpandLikeIdx() const {
+    return expand_like_idx_;
+  }
+
+  void SetupSequenceOperator(const workspace_t<Backend> &ws) {
+    expand_like_idx_ = -1;
+    auto num_inputs = ws.NumInput();
+    input_expand_desc_.clear();
+    input_expand_desc_.reserve(num_inputs);
+    for (int input_idx = 0; input_idx < num_inputs; input_idx++) {
+      const auto &input_shape = ws.GetInputShape(input_idx);
+      const auto &layout = GetInputLayout(ws, input_idx);
+      DALI_ENFORCE(layout.size() == 0 || layout.size() == input_shape.sample_dim(),
+                   make_string("Layout of input ", input_idx, " has size ", layout.size(),
+                               " which does not match the dimensionality of the input (got input "
+                               "with dimensionality ",
+                               input_shape.sample_dim(), ")."));
+      input_expand_desc_.emplace_back(input_shape, layout, ShouldExpandChannels(input_idx));
+      if (input_expand_desc_[input_idx].NumDimsToExpand() > 0) {
```

Comment:
```suggestion
      if (input_expand_desc_[input_idx].NumDimsToExpand() > 0 && expand_like_idx_ < 0) {
```
?

Reply:
Yeah, currently is does not matter really because all expandable inputs must be expandable in the same manner, so it could be last or first. If we were to allow for multi-inputs with different expandable layouts, like "FC" and "CF" at the same time, or "FC", "F", then it will need to be overridable I guess, as there's no way to tell what is it the proper layout.

I'll move setting the expand_like_idx_ to a separate method, so it is easier to make it overridable in the future.